include/nbl/builtin/hlsl/scan/virtual_workgroup.hlsl

```diff
@@ -6,87 +6,78 @@
+		const uint32_t levelInvocationIndex = localWorkgroupIndex * glsl::gl_WorkGroupSize().x + SubgroupContiguousIndex();
+		const bool lastInvocationInGroup = SubgroupContiguousIndex() == (gl_WorkGroupSize().x - 1);
 
+		const uint32_t lastLevel = params.topLevel << 1u;
+		const uint32_t pseudoLevel = treeLevel>params.topLevel ? (lastLevel-treeLevel):treeLevel;
 		const bool inRange = levelInvocationIndex <= params.lastElement[pseudoLevel];
 
-		Storage_t data = Binop::identity();
+        // REVIEW: Right now in order to support REDUCE operation we need to set the max treeLevel == topLevel
+        // so that it exits after reaching the top?
+
+		Storage_t data = Binop::identity(); // REVIEW: replace Storage_t with Binop::type_t?
 		if(inRange)
 		{
 			getData(data, levelInvocationIndex, localWorkgroupIndex, treeLevel, pseudoLevel);
 		}
 
 		if(treeLevel < params.topLevel) 
 		{
-			#error "Must also define some scratch accessor when calling operation()"
-			data = workgroup::reduction<Binop>()(data);
+            data = workgroup::reduction<Binop,ItemCount,device_capabilities>::template __call<Accessor>(data,accessor);
 		}
-		// REVIEW: missing _TYPE_ check and extra case here
+        else if (!isExclusive && params.topLevel == 0u)
+        {
+            data = workgroup::inclusive_scan<Binop,ItemCount,device_capabilities>::template __call<Accessor>(data,accessor);
+        }
 		else if (treeLevel != params.topLevel)
 		{
-			data = workgroup::inclusive_scan<Binop>()(data);
+			data = workgroup::inclusive_scan<Binop,ItemCount,device_capabilities>::template __call<Accessor>(data,accessor);
 		}
 		else
 		{
-			data = workgroup::exclusive_scan<Binop>()(data);
+			data = workgroup::exclusive_scan<Binop,ItemCount,device_capabilities>::template __call<Accessor>(data,accessor);
 		}
 		setData(data, levelInvocationIndex, localWorkgroupIndex, treeLevel, pseudoLevel, inRange);
 	}
-}
-}
-}
 
-#ifndef _NBL_HLSL_SCAN_MAIN_DEFINED_ // TODO REVIEW: Are these needed, can this logic be refactored?
-#include "nbl/builtin/hlsl/scan/default_scheduler.hlsl"
-namespace nbl
-{
-namespace hlsl
-{
-namespace scan
-{
 	DefaultSchedulerParameters_t getSchedulerParameters(); // this is defined in the final shader that assembles all the SCAN operation components
-	void main()
+	template<class Binop, typename Storage_t, bool isExclusive, uint16_t ItemCount, class Accessor>
+    void main(NBL_REF_ARG(Accessor) accessor)
 	{
 		const DefaultSchedulerParameters_t schedulerParams = getSchedulerParameters();
-		const uint topLevel = getParameters().topLevel;
+		const uint32_t topLevel = getParameters().topLevel;
 		// persistent workgroups
 		while (true)
 		{
-			uint treeLevel,localWorkgroupIndex;
-			if (scheduler::getWork(schedulerParams,topLevel,treeLevel,localWorkgroupIndex))
+            // REVIEW: Need to create accessor here.
+            // REVIEW: Regarding ItemsPerWG this must probably be calculated after each getWork call?
```

Comment:
ItemsPerWG should be a constant, your getData should pad inactive lanes with identity already